### PR TITLE
also learn schema name for MSSQL

### LINKED ISSUE #367: dm_from_src() MS SQL Server
Hi

I'm trying to import a database from our server, but get the error below.

`dw_dm <- dm_from_src(conn, learn_keys = TRUE)`

`Error: nanodbc/nanodbc.cpp:1617: 42000: [Microsoft][ODBC SQL Server Driver][SQL Server]Invalid object name 'AdditionalFundingRequestsFact'.  [Microsoft][ODBC SQL Server Driver][SQL Server]Statement(s) could not be prepared. 
<SQL> 'SELECT *
FROM AdditionalFundingRequestsFact AS "zzz15"
WHERE (0 = 1)'`

The table in question definitely exists (I can locate it and extract using `tbl() %>% collect()` just fine).
There are a number of different schema in the database, but it's the very first table in the database alphabetically.

Our server is running:
MSS version: 12.00.4416
R version: 4.0.0

### PR DESCRIPTION
**Pull request by TSchiefer**
now all tables can be accessed, independent of schema name.

In addition, IMHO an argument `schema` should also be provided for `dm_from_src()`.

closes #367

### CODE REVIEW

**Comment by TSchiefer** [comment]
Regarding ed3e7bb: for some reason we didn't realize earlier that `table_names` contained repetitive entries for each column in the respective tables. This didn't influence the end result though. (this has nothing to do with the intent of this PR, but I thought I'd add it, since it's at least somehow related)

**Comment by TSchiefer** [comment]
WIP cause I still have to ensure that the schema name is quoted correctly for MSSQL

File: R/dm-from-src.R [hunk 1]
```diff
@@ -38,7 +38,7 @@
 #' dm_from_src(con)
 #'
 #' DBI::dbDisconnect(con)
-dm_from_src <- function(src = NULL, table_names = NULL, learn_keys = NULL,
+dm_from_src <- function(src = NULL, table_names = NULL, learn_keys = NULL, schema = NULL,
```
Comment (krlmlr) [inline R/dm-from-src.R:41]: I'd rather leave out the formal argument for now, and pass it through the ellipsis the same way as with Postgres. This will make the pull request much smaller.

```suggestion
dm_from_src <- function(src = NULL, table_names = NULL, learn_keys = NULL,
```

File: R/learn.R [hunk 2]
```diff
@@ -119,9 +121,12 @@ mssql_learn_query <- function() { # taken directly from {datamodelr}
       and ind_col.column_id = cols.column_id
     left outer join sys.systypes [types] on
       types.xusertype = cols.system_type_id
+  where schema_name(tabs.schema_id) = %s
```
Comment (krlmlr) [inline R/learn.R:124]: Perhaps:

```suggestion
  where tabs.schema_id = schema_id(%s)
```

In the worst case it doesn't change anything, in the best case the query is a bit faster.

File: R/learn.R [hunk 3]
```diff
@@ -47,22 +47,23 @@ dm_learn_from_db <- function(dest, ...) {
...
   }
 
   table_names <-
     overview %>%
     arrange(table) %>%
-    select(schema, table) %>%
+    distinct(schema, table) %>%
     transmute(
       name = table,
       value = schema_if(schema, DBI::dbQuoteIdentifier(con, table))
```
Comment (krlmlr) [inline R/learn.R:69]: Suggesting to pass `con` to `schema_if()` and do the quoting there.

File: R/learn.R [hunk 4]
```diff
@@ -119,9 +121,12 @@ mssql_learn_query <- function() { # taken directly from {datamodelr}
       and ind_col.column_id = cols.column_id
     left outer join sys.systypes [types] on
       types.xusertype = cols.system_type_id
+  where schema_name(tabs.schema_id) = %s
   order by
     tabs.create_date,
-    cols.column_id"
+    cols.column_id",
+  dbQuoteString(con, schema)
```
Comment (krlmlr) [inline R/learn.R:128]:
```suggestion
    dbQuoteString(con, schema)
```

File: tests/testthat/test-dm-from-src.R [hunk 5]
```diff
@@ -8,7 +8,8 @@ test_that("table identifiers are quoted", {
     map_chr(dbplyr::remote_name)
 
   con <- dm_get_con(dm)
-  expect_equal(unname(remote_names), unclass(DBI::dbQuoteIdentifier(con, names(dm))))
+  # `gsub()`, cause schema names are part of the remote_names (also standard schemas "dbo" for MSSQL and "public" for Postgres).
+  expect_equal(gsub("^.*\\.", "", unname(remote_names)), unclass(DBI::dbQuoteIdentifier(con, names(dm))))
```
Comment (krlmlr) [inline tests/testthat/test-dm-from-src.R:12]: I wonder how this test could even succeed before?

Comment (TSchiefer) [reply tests/testthat/test-dm-from-src.R:12]: It does fail locally on Postgres for me. Anyway...

**Comment by TSchiefer** [comment]
regarding: af980c2

As I see it, in the current master `copy_dm_to()` is not ready to deal with schemas.
My problem when setting up a test with a `dm` in a schema was, that when I used something like:
```r
copy_dm_to(src_mssql, dm_for_disambiguate(), temporary = FALSE, table_names = function(x) {in_schema("testthat_for_dm", x)})
```
the tables would not end up in the schema "testthat_for_dm" but in "dbo", with names names like "testthat_for_dm.name_1" etc.

The problem is that the whole name was quoted as one object, but in order to deal with schemas, the dot between schema and table name can't be quoted.

I did not really fix this problem here, just made some sort of hack that assumes, that if the user already uses `ident_q`-class names, that they thought about this problematic and chose names that correctly address schemas.

I will open an issue addressing this though.

Meanwhile, if you don't like it, you can take it out, but then we need to also disable the tests until the issue is resolved (or you tell me how to use schemas in the current implementation of `copy_dm_to()`)

File: R/learn.R [hunk 6]
```diff
@@ -119,9 +122,12 @@ mssql_learn_query <- function() { # taken directly from {datamodelr}
       and ind_col.column_id = cols.column_id
     left outer join sys.systypes [types] on
       types.xusertype = cols.system_type_id
+  where tabs.schema_id = schema_id('%s')
```
Comment (krlmlr) [inline R/learn.R:125]:
```suggestion
  where tabs.schema_id = schema_id(%s)
```

File: R/learn.R [hunk 7]
```diff
@@ -119,9 +122,12 @@ mssql_learn_query <- function() { # taken directly from {datamodelr}
...
   order by
     tabs.create_date,
-    cols.column_id"
+    cols.column_id",
+    schema
```
Comment (krlmlr) [inline R/learn.R:129]:
```suggestion
    DBI::dbQuoteIdentifier(con, schema)
```

Otherwise the SQL is bad and prone to injection if `schema` contains `'`.

File: tests/testthat/test-learn.R [hunk 8]
```diff
@@ -1,49 +1,87 @@
...
-  def_original <-
-    dm_get_def(dm_for_filter()) %>%
-    select(-data)
-
-  expect_identical(
-    def_learned_renamed_reclassed,
-    def_original
+  # in case there happen to be other tables in schema "dbo"
+  dm_for_filter_mssql_learned <- dm_select_tbl(
+    dm_for_filter_mssql_learned_all,
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:19]: When using the pipe, please start with the first argument. Mixing piped and non-piped code is confusing.

```suggestion
  dm_for_filter_mssql_learned <-
    dm_for_filter_mssql_learned_all %>%
    dm_select_tbl(
```

File: tests/testthat/test-learn.R [hunk 9]
```diff
@@ -1,49 +1,87 @@
...
 
+  skip_if_src_not("mssql")
   # dm_learn_from_mssql() --------------------------------------------------
-  src_mssql <- skip_if_error(src_test("mssql"))
+  src_mssql <- my_test_src()
 
   # create an object on the MSSQL-DB that can be learned
   if (!any(src_tbls(src_mssql) %>%
     grepl("^tf_1_", .))) {
     copy_dm_to(src_mssql, dm_for_filter(), temporary = FALSE, table_names = unique_db_table_name)
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:12]: - Ensure qualified identifiers are not quoted
- Use copied dm to query actual table names for a more robust test

```suggestion
    dm_for_filter_copied <- copy_dm_to(src_mssql, dm_for_filter(), temporary = FALSE, table_names = ~ DBI::SQL(unique_db_table_name(.x)))
```

File: R/db-interface.R [hunk 10]
```diff
@@ -148,7 +148,9 @@ copy_dm_to <- function(dest, dm, ...,
         abort_copy_dm_to_table_names_duplicated(problem)
       }
 
-      table_names_out <- unclass(DBI::dbQuoteIdentifier(dest_con, table_names_out[src_names]))
+      if (!inherits(table_names_out,"ident_q")) {
```
Comment (krlmlr) [inline R/db-interface.R:151]: It already works if we pass in an already quoted object, e.g. created by `DBI::SQL()`.

File: tests/testthat/test-learn.R [hunk 11]
```diff
@@ -1,49 +1,87 @@
...
+    tf_1 = starts_with("tf_1"), tf_2 = starts_with("tf_2"), tf_3 = starts_with("tf_3"),
+    tf_4 = starts_with("tf_4"), tf_5 = starts_with("tf_5"), tf_6 = starts_with("tf_6"))
+
+  expect_equivalent_dm(
+    dm_for_filter_mssql_learned,
+    dm_for_filter()
+  )
+
+  walk(
+    dm_get_tables_impl(dm_for_filter_mssql_learned)[c(2, 1, 5, 6, 4, 3)],
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:31]: Should we index by name here instead?

File: tests/testthat/test-learn.R [hunk 12]
```diff
@@ -1,49 +1,87 @@
...
-
+test_that("Standard learning from MSSQL (schema 'dbo') works?", {
 
+  skip_if_src_not("mssql")
   # dm_learn_from_mssql() --------------------------------------------------
-  src_mssql <- skip_if_error(src_test("mssql"))
+  src_mssql <- my_test_src()
 
   # create an object on the MSSQL-DB that can be learned
   if (!any(src_tbls(src_mssql) %>%
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:10]: We should always `copy_dm_to()`.

File: tests/testthat/test-learn.R [hunk 13]
```diff
@@ -1,49 +1,87 @@
...
+  ) %>% dm_select_tbl(
+    tf_1 = starts_with("tf_1"), tf_2 = starts_with("tf_2"), tf_3 = starts_with("tf_3"),
+    tf_4 = starts_with("tf_4"), tf_5 = starts_with("tf_5"), tf_6 = starts_with("tf_6"))
+
+  expect_equivalent_dm(
+    dm_for_filter_mssql_learned,
+    dm_for_filter()
+  )
+
+  walk(
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:30]: This should be code in `on.exit()`, so that it is run also when the test aborts.

File: tests/testthat/test-learn.R [hunk 14]
```diff
@@ -1,49 +1,87 @@
...
+})
+
+
+test_that("Learning from specific schema on MSSQL works?", {
+
+  skip_if_src_not("mssql")
+  src_mssql <- my_test_src()
+  con_mssql <- src_mssql$con
+
+  # this schema name should be special enough to avoid any conflicts
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:43]: Can we use a randomized schema name?

File: R/learn.R (hunk 6 above)
Comment (TSchiefer) [reply R/learn.R:125]: I tried this and your next suggestion in all 3 combinations, but it won't work, not even with normal schema names. 
Probably the reason is that here we do not want to address the schema in a direct way, but we want to refer to the schema name representation in the result of the query. If we use `'%s'`, it is even possible to address schemas like "SELECT" or other keywords.

> Otherwise the SQL is bad and prone to injection if schema contains '.

Since I am not able to create a schema containing `'`, I unfortunately can't test this corner case.

File: R/learn.R (hunk 7 above)
Comment (TSchiefer) [reply R/learn.R:129]:
```suggestion
    gsub("'", "''", schema)
```
This way also schemas containing apostrophes are addressed correctly

File: R/db-interface.R (hunk 10 above)
Comment (krlmlr) [reply R/db-interface.R:151]: This means we no longer need this change?

File: R/learn.R (hunk 6 above)
Comment (krlmlr) [reply R/learn.R:125]: This is what `dbQuoteIdentifier()` is supposed to do.

File: R/learn.R (hunk 7 above)
Comment (krlmlr) [reply R/learn.R:129]: Maybe:

```suggestion
    unclass(DBI::dbQuoteIdentifier(con, schema))
```

? Let's not reimplement `dbQuoteIdentifier()` here.

Comment (TSchiefer) [reply R/learn.R:129]: this doesn't work: what we want for a schema called `SELECT` is `'SELECT'`, and with `unclass(DBI::dbQuoteIdentifier(con, schema))` we get `\"SELECT\"`.

for a schema with name `123fn'321` we get `\"123fn'321\"` from `unclass(DBI::dbQuoteIdentifier(con, "123fn'321"))`, but what we want is `'123fn''321'`.

what do you suggest?

File: R/db-interface.R (hunk 10 above)
Comment (TSchiefer) [reply R/db-interface.R:151]: ok, let's properly implement in #432 and until then do as you suggested

File: tests/testthat/test-learn.R (hunk 13 above)
Comment (TSchiefer) [reply tests/testthat/test-learn.R]: Good catch

**Comment by TSchiefer** [comment]
After 0e9d1b4 tests for MSSQL are passing when I let them run in the console with `testthis:::test_this()`.

But I get a strange behavior locally when I let the whole test-suite run with `devtools::test()`: I get warnings, but no errors for the tests for `dm_learn_from_src()`. The warnings indicate, that the keys could not be queried, and it tries to include all tables seen by `src_tbls(src)`, which of course doesn't work, cause there are some tables which are located in schemas. So I don't understand two things:

1. why doesn't the query work that focuses on a certain schema (or 'dbo')
1. if 1. happens, why is there only a warning and not a failing test? (the proper `dm` hasn't been learned)

It seems to work for the tic-run: `db-mssql / r-release` though.

File: R/learn.R (hunk 6 above)
Comment (krlmlr) [reply R/learn.R:125]: I mean `dbQuoteString()` -- you're right that `dbQuoteIdentifier()` is incorrect here.

File: R/learn.R [hunk 15]
```diff
@@ -119,9 +122,12 @@ mssql_learn_query <- function() { # taken directly from {datamodelr}
       and ind_col.column_id = cols.column_id
     left outer join sys.systypes [types] on
       types.xusertype = cols.system_type_id
+  where tabs.schema_id = schema_id('%s')
   order by
     tabs.create_date,
-    cols.column_id"
+    cols.column_id",
+  gsub("(^|[^'])(['])($|[^'])", "\\1\\2\\2\\3", schema)
```
Comment (krlmlr) [inline R/learn.R:129]:
```suggestion
  DBI::dbQuoteString(con, schema)
```

Comment (TSchiefer) [reply R/learn.R:129]: Ok, thanks, that looks way better :) Only thing is, that it also doubles `''`, so if the user tries to use their SQL knowledge, it will fail. But I think that's a minor issue.

File: tests/testthat/test-learn.R (hunk 13 above)
Comment (TSchiefer) [reply tests/testthat/test-learn.R]: FYI: I used `withr::defer()` in d07251a and the equivalent commit for the other test because I found this in the web: https://testthat.r-lib.org/articles/test-fixtures.html

File: tests/testthat/test-learn.R (hunk 14 above)
Comment (TSchiefer) [reply tests/testthat/test-learn.R]: would this be ok with you? 
715d93a and 1e835ab

**Comment by TSchiefer** [comment]
> Looks good. What does it take to unify the tests for MSSQL and Postgres? The CREATE SCHEMA and DROP SCHEMA syntax seems standardized.

seems to work now with 706c432

**Comment by TSchiefer** [comment]
maybe https://krlmlr.github.io/dm/articles/howto-dm-db.html#building-a-dm-from-a-subset-of-tables could be updated to include the possibility to learn from a schema on MSSQL and Postgres DBs? Here or in a future PR

File: tests/testthat/test-learn.R [hunk 16]
```diff
@@ -1,74 +1,86 @@
...
-
-  expect_identical(
-    def_learned_renamed_reclassed,
-    def_original
+  dm_for_filter_copied <- copy_dm_to(src_db, dm_for_filter(), temporary = FALSE, table_names = ~ DBI::SQL(unique_db_table_name(.x)))
+  order_of_deletion <- c("tf_2", "tf_1", "tf_5", "tf_6", "tf_4", "tf_3")
+  remote_tbl_names <- map_chr(
+    order_of_deletion,
+    ~dbplyr::remote_name(dm_for_filter_copied[[.x]])
+  ) %>% set_names(order_of_deletion)
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:25]:
```suggestion
    set_names(order_of_deletion),
    ~ dbplyr::remote_name(dm_for_filter_copied[[.x]])
  )
```

File: tests/testthat/test-learn.R [hunk 17]
```diff
@@ -1,74 +1,86 @@
...
+  dm_for_filter_db_learned_all <- dm_from_src(src_db)
+
+  # in case there happen to be other tables in schema "dbo" or "public"
+  dm_for_filter_db_learned <-
+    dm_for_filter_db_learned_all %>%
+    dm_select_tbl(!!!remote_tbl_names)
+
+  expect_equivalent_dm(
+    dm_for_filter_db_learned,
+    dm_for_filter()[order_of_deletion]
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:43]: Do we need to reorder?

```suggestion
    dm_for_filter()
```

File: tests/testthat/test-learn.R [hunk 18]
```diff
@@ -1,74 +1,86 @@
...
+    table_names = ~ DBI::SQL(dbplyr::in_schema(schema_name_q, .x))
   )
+  order_of_deletion <- c("iris_3", "iris_2", "iris_1")
+  remote_tbl_names <- set_names(paste0(schema_name_q, order_of_deletion), order_of_deletion)
+
+  withr::defer(
+    {
+      walk(
+        remote_tbl_names,
+        ~ dbExecute(con_db, paste0("DROP TABLE ", .x))
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:71]:
```suggestion
        ~ try(dbExecute(con_db, paste0("DROP TABLE ", .x)))
```

File: tests/testthat/test-learn.R [hunk 19]
```diff
@@ -1,74 +1,86 @@
...
+      )
+      dbExecute(con_db, paste0("DROP SCHEMA ", schema_name_q))
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:73]:
```suggestion
      try(dbExecute(con_db, paste0("DROP SCHEMA ", schema_name_q)))
```

File: tests/testthat/test-learn.R [hunk 20]
```diff
@@ -1,74 +1,86 @@
...
+  order_of_deletion <- c("tf_2", "tf_1", "tf_5", "tf_6", "tf_4", "tf_3")
+  remote_tbl_names <- map_chr(
+    order_of_deletion,
+    ~dbplyr::remote_name(dm_for_filter_copied[[.x]])
+  ) %>% set_names(order_of_deletion)
+
+  withr::defer(
+    walk(
+      dm_get_tables_impl(dm_for_filter_copied)[order_of_deletion],
+      ~ dbExecute(src_db$con, paste0("DROP TABLE ", dbplyr::remote_name(.x)))
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:30]:
```suggestion
      ~ try(dbExecute(src_db$con, paste0("DROP TABLE ", dbplyr::remote_name(.x))))
```

File: tests/testthat/test-learn.R [hunk 21]
```diff
@@ -1,74 +1,86 @@
...
+  )
+
+  dm_for_filter_db_learned <- dm_from_src(src_db, schema = schema_name) %>%
+    dm_select_tbl(!!!order_of_deletion)
 
   expect_equivalent_dm(
-    dm_postgres_learned_from_con_renamed,
-    dm_for_filter()
+    dm_for_filter_db_learned,
+    dm_for_disambiguate()[order_of_deletion]
```
Comment (krlmlr) [inline tests/testthat/test-learn.R:82]: Why is `dm_for_filter_...` compared with `dm_for_disambiguate()` ? Also:

```suggestion
    dm_for_disambiguate()
```

File: tests/testthat/test-learn.R (hunk 17 above)
Comment (TSchiefer) [reply tests/testthat/test-learn.R:43]: either reorder, or change the implementation of `expect_equivalent_dm()`; it fails currently, if the order of the tables in the `dm`-objects differ.

File: tests/testthat/test-learn.R (hunk 21 above)
Comment (TSchiefer) [reply tests/testthat/test-learn.R:82]: good catch :) I must have just copied & pasted (in the end it was a `dm_for_disambiguate` in disguise).
Don't know about reordering, see above.

**Comment by TSchiefer** [comment]
Think I addressed your points from the review apart from the reordering. Please feel free to solve this however you want or let me know.

**Comment by TSchiefer** [comment]
> I wonder if we should define a random seed for each test file?

I am not sure I understand how you'd like to implement that. Would the schema name then not be the same each time and therefore defy the reason for the function `random_schema()`?

**Comment by krlmlr** [comment]
Thanks!
